fabtests/prov/lpp/src/error.h

```diff
@@ -0,0 +1,143 @@
+		       my_node_name, ri->iteration, ##__VA_ARGS__)             \
+	} while (0)
+
+// Like error(), but with context about the current rank.
+#define ERROR(ri, fmt, ...)                                                    \
+	do {                                                                   \
+		debug_dump_trace((ri));                                        \
+		error("[rank:%ld node:%c iter:%ld] " fmt, ri->rank,            \
+		      my_node_name, ri->iteration, ##__VA_ARGS__)              \
+	} while (0)
+
+#define INSIST(ri, condition)                                                  \
+	do {                                                                   \
+		if (!(condition)) {                                            \
+			ERRORX(ri, "condition failed: " #condition);           \
+		}                                                              \
+	} while (0)
+
+#define INSIST_EQ(ri, value, testvalue, fmt)                                   \
+	do {                                                                   \
+		typeof(value) __cached_value = (value);                        \
+		typeof(value) __cached_testvalue = (testvalue);                \
+		if ((__cached_value) != (__cached_testvalue)) {                \
+			ERRORX(ri,                                             \
+			       "(" #value ") != (" #testvalue ")  (" fmt       \
+			       ") != (" fmt ")",                               \
+			       __cached_value, __cached_testvalue);            \
+		}                                                              \
+	} while (0)
+
+#define INSIST_FI_EQ(ri, value, testvalue)                                     \
+	do {                                                                   \
+		int insist_ret;                                                \
+		debug_trace_push((ri), __LINE__, __func__, __FILE__, #value);  \
+		if ((insist_ret = (value)) != (testvalue)) {                   \
+			ERRORX((ri),                                           \
+			       #value " == %d (fi_errno: %s), != " #testvalue  \
+				      " (%d)",                                 \
+			       insist_ret, fi_strerror(labs(insist_ret)),      \
+			       (testvalue));                                   \
+		}                                                              \
+		debug_trace_pop((ri));                                         \
+	} while (0)
+
+#define TRACE(ri, func)                                                        \
+	do {                                                                   \
+		debug_trace_push((ri), __LINE__, __func__, __FILE__, #func);   \
+		func;                                                          \
+		debug_trace_pop((ri));                                         \
+	} while (0)
+
+struct rank_info;
+void debug_trace_push(struct rank_info *ri, int line, const char *func,
+		      const char *file, const char *debugstr);
+void debug_trace_pop(struct rank_info *ri);
+void debug_dump_trace(struct rank_info *ri);
```

Comment:
libfabric defines common debug print capabilities for all providers.  See fi_log.h.  This makes it easier on users, since all providers follow the same logging format and environment controls, and allows apps to redirect log messages as needed.

If you want to propose extending that functionality with trace push/pop functionality, I would move that into separate patches.

Reply:
Yes, but this is in fabtests/prov/lpp/src for the lpp specific fabtest. Fabtests can not use FI_LOG macros.